netadapter/router/router.go

```diff
@@ -0,0 +1,112 @@
+package router
+
+import (
+	"github.com/kaspanet/kaspad/netadapter/id"
+	"github.com/kaspanet/kaspad/wire"
+	"github.com/pkg/errors"
+)
+
+// OnIDReceivedHandler is a function that is to be called
+// once a new Connection sends us its ID.
+type OnIDReceivedHandler func(id *id.ID)
+
+// OnRouteCapacityReachedHandler is a function that is to
+// be called when one of the routes reaches capacity.
+type OnRouteCapacityReachedHandler func()
+
+// Router routes messages by type to their respective
+// input channels
+type Router struct {
+	incomingRoutes map[string]*Route
+	outgoingRoute  *Route
+
+	onIDReceivedHandler           OnIDReceivedHandler
+	onRouteCapacityReachedHandler OnRouteCapacityReachedHandler
+}
+
+// NewRouter creates a new empty router
+func NewRouter() *Router {
+	router := Router{
+		incomingRoutes: make(map[string]*Route),
+		outgoingRoute:  NewRoute(),
+	}
+	router.outgoingRoute.setOnCapacityReachedHandler(func() {
+		router.onRouteCapacityReachedHandler()
+	})
+	return &router
+}
+
+// SetOnIDReceivedHandler sets the onIDReceivedHandler function for
+// this router
+func (r *Router) SetOnIDReceivedHandler(onIDReceivedHandler OnIDReceivedHandler) {
```

Comment:
All the OnIDReceived logic shouldn't pass through router - go directly to NetAdapter.